pipeline/mutate/mutator_id_token.go

```diff
@@ -152,7 +154,9 @@ func (a *MutatorIDToken) Mutate(r *http.Request, session *authn.AuthenticationSe
 		t := a.templates.Lookup(c.ClaimsTemplateID())
 		if t == nil {
 			var err error
+			a.templatesLock.Lock()
 			t, err = a.templates.New(c.ClaimsTemplateID()).Parse(c.Claims)
```

Comment:
The documentation says:

> Because associated templates share underlying data, template construction cannot be done safely in parallel. Once the templates are constructed, they can be executed in parallel.


So I think it would be better to have a `a.getTemplate(c.ClaimsTemplateID())` which looks up the template and if it doesn't exist creates it and adds it to the cache. The creation part would be locked while the read part would not have to be locked. This will increase throughput as we don't need to lock here for every request :)

Reply:
Hey, I'm not sure I understand. Do you have an example of what you're thinking?

Here are my assumptions right now with the current code:

```go
	var templateClaims []byte
	if len(c.Claims) > 0 {
		// find the already-constructed template:
		t := a.templates.Lookup(c.ClaimsTemplateID())
		if t == nil {
			// if we couldn't find it, lock and construct a new one (only one time)
			var err error
			a.templatesLock.Lock()
			t, err = a.templates.New(c.ClaimsTemplateID()).Parse(c.Claims)
			a.templatesLock.Unlock()
			if err != nil {
				return errors.Wrapf(err, `error parsing claims template in rule "%s"`, rl.GetID())
			}
		}

		// after a template already been constructed once, requests should reach this point without locking:
		var b bytes.Buffer
		if err := t.Execute(&b, session); err != nil {
			return errors.Wrapf(err, `error executing claims template in rule "%s"`, rl.GetID())
		}

		templateClaims = b.Bytes()
		if err := json.NewDecoder(&b).Decode(&claims); err != nil {
			return errors.WithStack(err)
		}
	}
```

Thanks!

Reply:
Sorry, my bad, I just overlooked the preceding if statement and template lookup - you are correct, this is indeed the way it can be solved! :)